### ISSUE TITLE
Inconsistent output shape for InnerProduct

### ISSUE THREAD

**Issue by csukuangfj** [Tencent/ncnn#4728]
The output shape of `InnerProduct` depends on  `input_shape.h`.
For instance, if `input_shape.dim` is 2 and `input_shape.h=1`, then the output.dim becomes 1. But we want
`output_shape.dim == input_shape.dim`.

Reproducible with the latest master as of today.

# Code for reproducing
```python3
#!/usr/bin/env python3

import os

import ncnn
import torch
import torch.nn as nn


class Foo(nn.Module):
    def __init__(
        self,
    ):
        super().__init__()
        self.linear = nn.Linear(2, 5)

    def forward(self, x):
        return self.linear(x)


@torch.no_grad()
def main():
    f = Foo()
    f.eval()
    x = torch.tensor([1, 2.0])
    m = torch.jit.trace(f, x)
    m.save("m.pt")
    os.system("pnnx ./m.pt")

    param = "m.ncnn.param"
    model = "m.ncnn.bin"
    with ncnn.Net() as net:
        net.load_param(param)
        net.load_model(model)

        x0 = torch.rand(1, 2)
        x1 = torch.rand(2, 2)

        with net.create_extractor() as ex:
            ex.input("in0", ncnn.Mat(x0.numpy()).clone())
            ret, y0 = ex.extract("out0")
            y0 = torch.from_numpy(y0.numpy()).clone()

        with net.create_extractor() as ex:
            ex.input("in0", ncnn.Mat(x1.numpy()).clone())
            ret, y1 = ex.extract("out0")
            y1 = torch.from_numpy(y1.numpy()).clone()

        print(y0.shape)  # torch.Size([5])
        print(y1.shape)  # torch.Size([2, 5])


if __name__ == "__main__":
    main()
```

# Expected output

The shape of `y0` should be `torch.Size([1, 5])`.

# Possible fixes
https://github.com/Tencent/ncnn/blob/d4046b4ae90b905bd2b292f008af07a04756cff5/src/layer/innerproduct.cpp#L118

Remove `h > 1` in the above statement

# Why do I need it

I need it for [sherpa-ncnn](https://github.com/k2-fsa/sherpa-ncnn)

We have the following code from our PyTorch model
https://github.com/k2-fsa/icefall/blob/bccd20d9786e2ec5f1445e1b20fb806ddfff78b1/egs/librispeech/ASR/lstm_transducer_stateless/lstm.py#L600-L601
```python3
            x = x.permute(0, 2, 1, 3).reshape(1, -1, self.conv_out_dim)
            x = self.out(x)
```
where `self.out` is of type `nn.Linear`.  The issue is that in ncnn, the output of `self.out(x)` becomes a 1-D tensor, which becomes problematic for my later operators.

I have modified ncnn in my own fork. See https://github.com/csukuangfj/ncnn/pull/2/files

But now I want to use ncnn directly without any modifications and the above problem becomes an issue for 
[sherpa-ncnn](https://github.com/k2-fsa/sherpa-ncnn)

**Comment by nihui** [comment]
https://github.com/Tencent/ncnn/pull/4730

**Comment by nihui** [comment]
resolved
